### ISSUE TITLE
AutoMapper unable to cast FakeDbAsyncEnumerable to IQueryable

### ISSUE THREAD

**Issue by dhartjes** [ReversePOCO/EntityFramework-Reverse-POCO-Code-First-Generator#390]
Thanks so much for the awesome tool!

Similar to in this StackOverflow [item](https://stackoverflow.com/questions/44807618/automapper-unable-to-cast-testdbasyncenumerable-to-iqueryable), I am not able to use AutoMapper's ProjectTo when using the generated FakeDbContext, but it works fine with the real one. The fixes proposed in StackOverflow resolve this issue when I place them in my generated context class, but will of course be overwritten upon future updates to the .tt file.

I thought perhaps there might be some way to configure the Reverse-POCO-Generator to make the FakeDbAsyncQueryProvider<TEntity> class a partial and then include these changes somewhere they won't get overwritten, but this does not seem to be the case.

I did a performance test with and without the changes, and there was no discernible performance difference. Are there any potential downsides?

Thanks!
Dominic

Changes:

```c#
   [System.CodeDom.Compiler.GeneratedCode("EF.Reverse.POCO.Generator", "2.33.0.0")]
   public class FakeDbAsyncQueryProvider<TEntity> : System.Data.Entity.Infrastructure.IDbAsyncQueryProvider
   {
      private readonly IQueryProvider _inner;

      public FakeDbAsyncQueryProvider(IQueryProvider inner)
      {
         _inner = inner;
      }

      public IQueryable CreateQuery(System.Linq.Expressions.Expression expression)
      {
         //return new FakeDbAsyncEnumerable<TEntity>(expression);

         switch (expression)
         {
            case System.Linq.Expressions.MethodCallExpression m:
            {
               var resultType = m.Method.ReturnType; // it shoud be IQueryable<T>
               var tElement = resultType.GetGenericArguments()[0];
               var queryType = typeof(FakeDbAsyncEnumerable<>).MakeGenericType(tElement);
               return (IQueryable)System.Activator.CreateInstance(queryType, expression);
            }
         }
         return new FakeDbAsyncEnumerable<TEntity>(expression);
      }

      public IQueryable<TElement> CreateQuery<TElement>(System.Linq.Expressions.Expression expression)
      {
         //return new FakeDbAsyncEnumerable<TElement>(expression);

         var queryType = typeof(FakeDbAsyncEnumerable<>).MakeGenericType(typeof(TElement));
         return (IQueryable<TElement>)System.Activator.CreateInstance(queryType, expression);
      }


```

**Comment by sjh37** [comment]
Thanks very much for this Dominic.
Added to the project and will be in the next release :-)

**Comment by dhartjes** [comment]
Exciting! Thanks a lot!

**Comment by sjh37** [comment]
Released in v2.37.0

**Comment by dhartjes** [comment]
Exciting! Thanks again for the supremely useful tool!
